cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,336 @@
+        break;
+
+      case GeometryType::LINESTRING: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_coords, src->ReadUInt32(swap));
+        ARROW_RETURN_NOT_OK(
+            ReadSequence(src, geometry_type_and_dimensions.second, n_coords, swap));
+        break;
+      }
+      case GeometryType::POLYGON: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_parts, src->ReadUInt32(swap));
+        for (uint32_t i = 0; i < n_parts; i++) {
+          ARROW_ASSIGN_OR_RAISE(uint32_t n_coords, src->ReadUInt32(swap));
+          ARROW_RETURN_NOT_OK(
+              ReadSequence(src, geometry_type_and_dimensions.second, n_coords, swap));
+        }
+        break;
+      }
+
+      // These are all encoded the same in WKB, even though this encoding would
+      // allow for parts to be of a different geometry type or different dimensions.
+      // For the purposes of bounding, this does not cause us problems.
+      case GeometryType::MULTIPOINT:
+      case GeometryType::MULTILINESTRING:
+      case GeometryType::MULTIPOLYGON:
+      case GeometryType::GEOMETRYCOLLECTION: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_parts, src->ReadUInt32(swap));
+        for (uint32_t i = 0; i < n_parts; i++) {
+          ARROW_RETURN_NOT_OK(ReadGeometryInternal(src, /*record_wkb_type*/ false));
+        }
+        break;
+      }
+    }
+
+    return ::arrow::Status::OK();
+  }
+
+  ::arrow::Status ReadSequence(WKBBuffer* src, Dimensions dimensions, uint32_t n_coords,
+                               bool swap) {
+    switch (dimensions) {
+      case Dimensions::XY:
+        return src->ReadDoubles<BoundingBox::XY>(
+            n_coords, swap, [&](BoundingBox::XY coord) { box_.UpdateXY(coord); });
+      case Dimensions::XYZ:
+        return src->ReadDoubles<BoundingBox::XYZ>(
+            n_coords, swap, [&](BoundingBox::XYZ coord) { box_.UpdateXYZ(coord); });
+      case Dimensions::XYM:
+        return src->ReadDoubles<BoundingBox::XYM>(
+            n_coords, swap, [&](BoundingBox::XYM coord) { box_.UpdateXYM(coord); });
+      case Dimensions::XYZM:
+        return src->ReadDoubles<BoundingBox::XYZM>(
+            n_coords, swap, [&](BoundingBox::XYZM coord) { box_.UpdateXYZM(coord); });
+      default:
+        return ::arrow::Status::Invalid("Unknown dimensions");
+    }
+  }
+
+  static ::arrow::Result<GeometryTypeAndDimensions> ParseGeometryType(
+      uint32_t wkb_geometry_type) {
+    // The number 1000 can be used because WKB geometry types are constructed
+    // on purpose such that this relationship is true (e.g., LINESTRING ZM maps
+    // to 3002).
+    uint32_t geometry_type_component = wkb_geometry_type % 1000;
+    uint32_t dimensions_component = wkb_geometry_type / 1000;
+
+    auto min_geometry_type_value = static_cast<uint32_t>(GeometryType::POINT);
+    auto max_geometry_type_value =
+        static_cast<uint32_t>(GeometryType::GEOMETRYCOLLECTION);
+    auto min_dimension_value = static_cast<uint32_t>(Dimensions::XY);
```

Comment:
it might be cleaner to have a specific Min/max member of MIN/MAX (I believe you can have two symbols pointing the same value).